Enforce `-` as word separator in devicetree property names

mbolivar-nordic opened the pull request:
Fixes: #53506

This tree-wide change cleans up existing property names which use `_` instead of `-`, and adds a CI check to avoid new ones from creeping in. This requires a little bit of prep work to add a helper function to edtlib to make the check easier to write.

I found a few issues in check_compliance.py while I was adding this that I took the time to fix while I was here.

mbolivar-nordic committed:
treewide: use dashes in DTS properties, not underscores

The canonical way to separate words in the name of a devicetree
property is by using a dash (-), not an underscore (_).

Clean up nonconforming bindings and DTS files in tree so we are
consistent about this upstream.

This patch should have no impact on source code, since dashes get
converted to underscores when consuming values from
<zephyr/devicetree.h>.

However, out of tree DTS files using the affected bindings will need
to be updated to match the new names.

For the record, I tried to see if any of these properties are coming
from somewhere in upstream Linux's Documentation/devicetree/bindings,
and I don't think they are:

- ftdi,ft800: nothing relevant found

- litex,i2s: nothing relevant found

- microchip,xec-*: nothing relevant found

- microchip,mcp4728: iio/dac/microchip,mcp4725.yaml seems like the
  best match; it doesn't mention these properties

- nuvoton,nct38xx-gpio-port: nothing relevant found

- nxp,imx*: nothing relevant found

- nxp,mcux-edma: nothing relevant found

- silabs,gecko-ethernet: nothing relevant found

- snps,creg-gpio: none of these are mentioned in
  gpio/snps,creg-gpio.txt, and drivers/gpio/gpio-creg-snps.c seems to be
  hard-coding the values. Looks like zephyr went its own
  way here.

- solomon,ssd1306fb-spi.yaml: display/solomon,ssd1307fb.yaml seems
  like the best match; it doesn't mention these properties

- ti,lp503x: leds/leds-lp50xx.yaml seems like the best match;
  it doesn't mention these properties

Note that the mmc-hs200-1_8v and mmc-hs400-1_8v property names defined
in sdhc.yaml exist in Linux, so we are OK with those.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
gen_devicetree_rest: prep work refactor

Move some code around in preparation to expose a bit of functionality
more widely from within edtlib itself.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
edtlib: add bindings_from_dirs

Add a helper for getting all the bindings under some iterable of
directories. This is currently done in the gen_devicetree_rest script,
and is useful enough to expose as a package-level helper.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
gen_devicetree_rest: use edtlib.bindings_from_dirs

Rely on the helper in the library now that it's available,
for better separation of concerns.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
edtlib: move an internal helper to a better place

The _init_graph() helper is called from __init__() after _init_nodes()
and before _init_luts(). Move its definition in between the
definitions of the others accordingly, to keep things tidy.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
check_compliance: fix path_hint behavior

As far as I can tell, this bug has existed since the script was first
introduced into the main tree in
288ae28c13c85d44772543894784682eb8cc8d63.

It's not clear from that commit's message where this code came from
originally, so I can't tell how far back the bug goes before that.

Each ComplianceTest must declare a path_hint class attribute. The test
framework claims it will run the test from that directory. However,
this behavior doesn't actually seem to be supported, and the test is
just run from wherever the script is called from.

I think maybe nobody noticed until now for two reasons:

1. compliance.yml happens to run the script from ZEPHYR_BASE, which
   is what the tests mostly seem to want anyway

2. the test run() methods seem to make no assumptions about where they
   are run from

Either way this is confusing and wrong; fix it. This will enable some
cleanup patches too.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
check_compliance: rely on path_hint + ZEPHYR_BASE behavior

Now that we've fixed the path_hint feature to work as specified and
we've ensured that ZEPHYR_BASE is always set, we can rely on these
facts from the test case code. This lets us simplify various things,
improving readability.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
check_compliance: provide a default ZEPHYR_BASE

This script is in the zephyr repository; there's no reason not to have
a reasonable default for this variable if the user hasn't overridden
it.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
check_compliance: re-work DevicetreeBindingsCheck

Use the edtlib API instead of parsing bindings by hand. This allows us
to avoid false positives (e.g. if 'required: false' appears in a
description string) and make sure we are only checking the YAML syntax
that makes up bindings themselves.

As an optimization, don't perform checks when the commit range hasn't
changed any bindings. This saves the several seconds it takes to parse
all the bindings files when that would be wasted effort.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
edtlib: refactor warning handling in EDT

Factoring out the decision making for how to handle warnings into a
common routine will lead to less boilerplate when adding new warnings.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

fabiobaltieri commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -538,6 +539,33 @@ def _check(self):
+            return
+
+        # The regular expression comes from dt-schema.
+        compat_re = r'^[a-zA-Z][a-zA-Z0-9,+\-._]+$'
```
Should we stick to lowercase? That's what I was going on in my version of this, it matches what we have right now in the repository, would save some future incoherence with capitalization.

mbolivar-nordic replied:
I don't think so. The regular expression we are using here is the one that dt-schema itself is using to validate compatibles. Why should we be stricter?

fabiobaltieri replied:
Because that is the style that we effectively adopted right now, the current set of compatibles (all 1200+ of them) are coherent in being all lowercase and the vendor names are all lowercase. It'd be nice to stick to it, avoid a code review conversation in the line of "but the standard allows it" versus "all the others are coherent".

mbolivar-nordic replied:
I checked upstream Linux and there are compatibles which use uppercase letters across multiple vendors (including big names like samsung, ti, fsl, brcm, intel, national, marvell). Given that, do you still think this is a good idea?

fabiobaltieri replied:
Yeah I checked as well and honestly I find them to look odd, for example `"atmel,at25256B"`. I would find it nice to be strict and just stick to lowercase since we landed there naturally. Maybe seek for a third opinion? @galak?

fabiobaltieri replied:
Another thought: if we try and be strict now and then it turns out we change our mind, it's easy to change the check to be less strict. Not so much the other way around.

mbolivar-nordic replied:
OK, I guess we can start there for now, but I'm going to resolve this comment and move the compatibles check to another PR now that we've got the tree-wide `s/_/-/g` change approved.

mbolivar-nordic committed:
check_compliance: check for underscores in property names

Issue failures when a devicetree property defined in a binding
contains an underscore. The standard word separator in a property is a
dash.

(This won't affect the standard -- albeit deprecated -- device_type
property, because that is not defined in any bindings.)

Add an 'allowlist' type file, check_compliance_ok_properties.yaml,
for existing cases that are OK.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/53506
Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented on scripts/ci/check_compliance.py:
```diff
@@ -704,6 +708,38 @@ class KconfigBasicCheck(KconfigCheck):
+    path_hint = ZEPHYR_BASE
+
+    def run(self):
+        bindings_dir = str(Path(ZEPHYR_BASE) / 'dts' / 'bindings')
```
> There's already a DevicetreeBindingsCheck calss here that goes over all the binding files and does few checks, can this integrate into that? 

Of course, thanks for catching that :). I botched the rebase on that part. Fixed.

> My main thought is that this filer is only checking the bindings in $ZEPHYR_BASE/dts/bindings, but the buildsystem allows extra bindings on the application directory as well and it would be cool if we could check those too.

Agree, it would be nice, but I don't think the current checks are doing that; are they?

At any rate, if we can get a list of directories to check from the calling environment, we can extend the current code to change `edtlib.bindings_from_dirs([bindings_dir])` to `edtlib.bindings_from_dirs(list_of_bindings_dirs)` very easily.

Can you suggest a way to get the list we need into that method?

fabiobaltieri replied:
> > My main thought is that this filer is only checking the bindings in $ZEPHYR_BASE/dts/bindings, but the buildsystem allows extra bindings on the application directory as well and it would be cool if we could check those too.
> 
> Agree, it would be nice, but I don't think the current checks are doing that; are they?

They are right now, check is:

https://github.com/zephyrproject-rtos/zephyr/blob/f09111a2a97aa3457d2a6cf9c362c53133348d3e/scripts/ci/check_compliance.py#L245

so it looks for anything with `dts/bindings` in the path, the buildsystem code is somewhat similar:

https://github.com/zephyrproject-rtos/zephyr/blob/f09111a2a97aa3457d2a6cf9c362c53133348d3e/cmake/modules/dts.cmake#L54-L60

https://github.com/zephyrproject-rtos/zephyr/blob/f09111a2a97aa3457d2a6cf9c362c53133348d3e/cmake/modules/dts.cmake#L141

> At any rate, if we can get a list of directories to check from the calling environment, we can extend the current code to change `edtlib.bindings_from_dirs([bindings_dir])` to `edtlib.bindings_from_dirs(list_of_bindings_dirs)` very easily.
> 
> Can you suggest a way to get the list we need into that method?

I don't think we can cover all cases since one may come up with some weird combination of binding inclusions from multiple roots or custom paths, but I think for the scope of this it may be good enough to load the base bindings + a potential root of any file modified by the patch currently being tested. So we could use the `get_files(filter="d")` list to find all the potential roots (anything under a 'dts/bindings' sub path) and then do the check on "base" and any "base + external root".

fabiobaltieri replied:
There's few bindings in the main repository outside of `$ZEPHYR_BASE/dts/bindings`

```
$ find -type f | grep dts/bindings | grep -v '\./dts/bindings/'
./boards/arm/arty/dts/bindings/arm,daplink-qspi-mux.yaml
./boards/arm/efr32bg_sltb010a/dts/bindings/silabs,gecko-wake-up-triggers.yaml
./boards/arm/twr_ke18f/dts/bindings/nxp,flexio.yaml
./boards/arm/actinius_icarus/dts/bindings/actinius-charger-enable.yaml
./boards/arm/actinius_icarus/dts/bindings/actinius-sim-select.yaml
./boards/arm/pinetime_devkit0/dts/bindings/misc/pine64,pinetime-key-out.yaml
./boards/arm/96b_wistrio/dts/bindings/qorvo,rfsw8001.yaml
...
```
And more in `samples` and `tests`.

The current code is considering those as well -- but I'm also interested in having this check usable in an external repository (would like to run it as part of our EC codebase presubmit), so it'd be ideal if it would just work.

mbolivar-nordic replied:
> The current code is considering those as well

I restored this functionality; thanks.

> I'm also interested in having this check usable in an external repository

Fair enough, but that's out of scope for what I'm trying to do since the current code isn't doing that.

mbolivar-nordic commented:
@de-nordic I need https://github.com/zephyrproject-rtos/fatfs/pull/16 for this PR

henrikbrixandersen commented on boards/arm/mec1501modular_assy6885/mec1501modular_assy6885.dts:
```diff
@@ -69,8 +69,7 @@
 };
 
 &i2c_smb_0 {
-	status = "okay";
```
Was this removal intentional?

henrikbrixandersen commented on boards/arm/mec15xxevb_assy6853/mec15xxevb_assy6853.dts:
```diff
@@ -131,8 +131,7 @@
 };
 
 &i2c_smb_2 {
-	status = "okay";
```
Same here?

henrikbrixandersen reviewed (changes requested):
Looks good, couple of nits:

fabiobaltieri commented on scripts/ci/check_compliance.py:
```diff
@@ -34,6 +35,21 @@
 
+@contextlib.contextmanager
+def run_in_directory(dir):
+    # Helper to run a code block from a given directory.
+    # Use it like this:
+    #
+    #     with run_in_directory('some/directory'):
+    #         do_something()
```
nit: can you make this pydoc comment instead?

fabiobaltieri reviewed (commented):
Cool, thanks Martí. Thinking we could add some more strict test as a start but this looks like a solid foundation to build up upon.

de-nordic commented:
@henrikbrixandersen Sorry for calling you to the re-review, somehow I managed to press the button in wrong review.

mbolivar-nordic commented:
Previous force push is a rebase and removal of the west.yml patch

mbolivar-nordic commented:
@henrikbrixandersen thank you for catching the `status = "okay";` removals; those were artifacts of a botched rebase. Fixed.

henrikbrixandersen commented:
@mbolivar-nordic At least the compliance check works 😂

carlescufi commented on scripts/ci/check_compliance.py:
```diff
@@ -25,8 +25,15 @@
+
 logger = None
 
+# This ends up using the current repository if ZEPHYR_BASE is unset.
```
But we already have a default for `ZEPHYR_BASE`:

https://github.com/zephyrproject-rtos/zephyr/blob/ee9903005a6cebdce4c0cc5b8bf8144c9ec8ab7c/scripts/ci/check_compliance.py#L1110-L1118

mbolivar-nordic replied:
> But we already have a default for ZEPHYR_BASE:

Oh, I didn't see that this function was changing global variables. That's fun.

I still need this at module level so that `check_compliance: re-work DevicetreeBindingsCheck` later on in the PR works, so I'll just refactor this patch to remove the lines you mention as well.

carlescufi replied:
But you did not remove those lines, did you?

carlescufi commented:
You wrote:

> 1. compliance.yml happens to run the script from ZEPHYR_BASE, which
   is what the tests mostly seem to want anyway

But that's only true upstream. Downstream this is run from the root of the repo that the checks need to be run on.

mbolivar-nordic commented:
> But that's only true upstream. Downstream this is run from the root of the repo that the checks need to be run on.

You're saying that we are running the zephyr/.github/workflows/compliance.yml file from non-zephyr repositories in downstreams?

If so, I'm having trouble understanding if you're asking for a change, though. The documented `path_hint` behavior is

```
    path_hint:
      The path the test runs itself in.  This is just informative and used in
      the message that gets printed when running the test.
```

and this PR is making the implementation match this documentation, so I don't see any changes that need doing, except maybe removing the "just informative" part. Am I missing something?

mbolivar-nordic commented:
@carlescufi ping, can you please explain what the changes you are requesting are? See my previous comment.

fabiobaltieri commented:
@carlescufi ping? It'd be cool to move forward with this.

henrikbrixandersen commented:
This needs rebasing.

cfriedt commented:
@mbolivar-nordic - please rebase and resolve conflicts. I'll approve again

henrikbrixandersen commented:
Removing the `stale` label as I hope this will be picked up again.

carlescufi commented:
@mbolivar-nordic apologies for the huge delay. Your patches make sense, in general, I will add a comment in the code itself.

carlescufi commented:
> You're saying that we are running the zephyr/.github/workflows/compliance.yml file from non-zephyr repositories in downstreams?

No, not the same exact workflow, but a very similar one. See here:
https://github.com/nrfconnect/sdk-nrf/blob/7b40e24932178c3686fcd3dee71796b7b7962dc7/.github/workflows/compliance.yml#L78

cfriedt commented:
This would be a good thing to redo before lts3

henrikbrixandersen commented:
> This would be a good thing to redo before lts3

Agreed. Would love to see this happen.
